dvc/repo/__init__.py

```diff
@@ -307,6 +308,11 @@ def graph(self, stages=None, from_directory=None):
                 if path_dir.startswith(out.path + os.sep):
                     raise StagePathAsOutputError(stage.wdir, stage.relpath)
 
+        for index, out in enumerate(outs):
+            for o in outs[index:]:
```

Comment:
Why do we only start from `index` though? Could you explain please?

Reply:
It is added in order to reduce number of checks.
We have two options that should trigger exception raise:
1. When `out.path.startswith(o.path + os.sep)`
2. When `o.path.startswith(out.path + os.sep)`
Because I check both cases in one if, there is no need to go back to checking out if it has been already compared to other outs. So index stands as guard that prevents performing same check twice.

Reply:
@pared Great point! Could add a comment about it, please?

Reply:
Obsolete, after priv discussion overlapping check moved to other loop.